### PR TITLE
Propagate Debugger… attributes from async/iterator method to the corresponding state machine MoveNext method.

### PR DESCRIPTION
**Pull request by AlekseyTs**
This addresses issue #807.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@gafter, @VSadov, @jaredpar, @agocke Please review.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 1]
```diff
@@ -1255,17 +1255,16 @@ private BoundStatement ChainImplicitStructConstructor(MethodSymbol methodSymbol,
                 var codeGen = new CodeGen.CodeGenerator(method, block, builder, moduleBuilder, diagnosticsForThisMethod, optimizations, generateDebugInfo);
 
                 // We need to save additional debugging information for MoveNext of an async state machine.
-                var stateMachineMethod = method as SynthesizedStateMachineMethod;
-                bool isStateMachineMoveNextMethod = stateMachineMethod != null && method.Name == WellKnownMemberNames.MoveNextMethodName;
+                var stateMachineMoveNextMethod = method as SynthesizedStateMachineMoveNextMethod;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: This change makes the code more fragile (tight coupling, by making the code sensitive to irrelevant refactoring of implementation types). If you want a "more precise" test than was there before, please add a virtual boolean property to SynthesizedStateMachineMethod or some supertype and test that property.

File: src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs [hunk 2]
```diff
@@ -137,34 +137,28 @@ private void GenerateEnumeratorImplementation()
...
-                    debuggerHidden: IsDebuggerHidden(this.method),
-                    hasMethodBodyDependency: true);
+                var moveNextMethod = OpenMoveNextMethodImplementation(IEnumerator_MoveNext);
 
                 GenerateMoveNextAndDispose(moveNextMethod, disposeMethod);
             }
 
             // Add T IEnumerator<T>.Current
             {
-                OpenPropertyImplementation(IEnumeratorOfElementType_get_Current, debuggerHidden: true, hasMethodBodyDependency: false);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs]: It isn't clear to me how the debuggerHidden behavior is preserved from the previous revision of the code in the absence of the scenario this PR is intended to change.

File: src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs [hunk 3]
```diff
@@ -137,34 +137,28 @@ private void GenerateEnumeratorImplementation()
...
+                OpenPropertyImplementation(IEnumeratorOfElementType_get_Current);
                 F.CloseMethod(F.Return(F.Field(F.This(), _currentField)));
             }
 
             // Add void IEnumerator.Reset()
             {
-                OpenMethodImplementation(IEnumerator_Reset, debuggerHidden: true, generateDebugInfo: false, hasMethodBodyDependency: false);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs]: Ditto, and in other places.

File: src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs]: Oh, I see, it is always "true" except in the new subclass you've created. Looks good.

**Comment by tmat** [comment]
:+1:

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/SynthesizedStateMachineMethod.cs [hunk 4]
```diff
@@ -9,19 +10,18 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// <summary>
     /// State machine interface method implementation.
     /// </summary>
-    internal sealed class SynthesizedStateMachineMethod : SynthesizedImplementationMethod, ISynthesizedMethodBodyImplementationSymbol
+    internal abstract class SynthesizedStateMachineMethodBase : SynthesizedImplementationMethod, ISynthesizedMethodBodyImplementationSymbol
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/SynthesizedStateMachineMethod.cs]: This is a poor name; it represents all synthesized state machines, so it should be called `SynthesizedStateMachineMethod`. The more specialized subclasses can have more specific names, but we should avoid having the code's behavior based on the shape of the hierarchy of implementation classes.

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 5]
```diff
@@ -1493,10 +1493,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 Dim codeGen = New codeGen.CodeGenerator(method, block, builder, moduleBuilder, diagnostics, optimizations, generateDebugInfo)
 
                 ' We need to save additional debugging information for MoveNext of an async state machine.
-                Dim stateMachineMethod = TryCast(method, SynthesizedStateMachineMethod)
+                Dim stateMachineMoveNextMethod = TryCast(method, SynthesizedStateMachineMoveNextMethod)
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: I'm not as concerned about this pattern in the VB compiler as it is already very widely used.

**Comment by gafter** [comment]
Other than two things in the C# code (renaming of `SynthesizedStateMachineMethod` to `SynthesizedStateMachineMethodBase` and a type test to the implementation type `SynthesizedStateMachineMoveNextMethod`), looks good.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: This change doesn't make the code more fragile than it was before. The type name change reflects the change in type hierarchy. Please open a separate issue if you would like the type test removed from this code altogether.

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/SynthesizedStateMachineMethod.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/SynthesizedStateMachineMethod.cs]: @gafter, Please suggest a better names for the affected types and I will be happy to use them.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: In the C# code base we avoid type tests except to types that are internal APIs, which is usually evidenced by the fact that they introduce members that we need to use. The existing type test was to a type that did indeed have such members (e.g. `stateMachineMethod.StateMachineType`). That is why I suggested adding a virtual boolean property in `StateMachineMethod` for this purpose, though I think the previous version of the code that just used the Name property was appropriate.

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/SynthesizedStateMachineMethod.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/SynthesizedStateMachineMethod.cs]: I suggest `SynthesizedStateMachineMethod` for the common (abstract) base, `SynthesizedStateMachineMoveNextMethod` and `SynthesizedStateMachineDebuggerHiddenMethod` for the two concrete classes. In that case you could either continue to test the .Name property, or add a virtual property `IsMoveNext` for the places where you need that test.

Alternately, if you're really ambitious, you could move some of the dependent behavior into virtual methods, but that seems far out of scope for this PR.

Comment (tmat) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/SynthesizedStateMachineMethod.cs]: Why ```SynthesizedStateMachineDebuggerHiddenMethod```?

**Pull request merged**